src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -145,20 +154,33 @@ public void actionPerformed(@NotNull AnActionEvent e) {
       return;
     }
 
-    startAutoInstallsWithRestart(course, !courseProjectViewModel.userOptsOutOfSettings(), project);
+    boolean importIdeSettings = !courseProjectViewModel.userOptsOutOfSettings();
 
-    if (!tryImportProjectSettings(project, course)) {
-      return;
-    }
+    Future<?> autoInstallDone = executor.submit(() -> startAutoInstalls(course, project));
 
-    if (!courseProjectViewModel.userOptsOutOfSettings() && !tryImportIdeSettings(project, course)) {
-      return;
-    }
+    Future<Boolean> projectSettingsImported =
+        executor.submit(() -> tryImportProjectSettings(project, course));
+
+    Future<Boolean> ideSettingsImported =
+        executor.submit(() -> importIdeSettings && tryImportIdeSettings(project, course));
```

Comment:
`SettingsImporter` should probably now use read actions where needed as it gets run on a background thread.

Also, I assume you've reasoned through that importing project settings and IDE settings in parallel is safe? (Can't see a reason why it wouldn't be)

Reply:
> SettingsImporter should probably now use read actions where needed as it gets run on a background thread.

A good point but I didn't find anything that should be write-locked (that is, any PSI, VFS, or project root model things).

> Also, I assume you've reasoned through that importing project settings and IDE settings in parallel is safe?

I don't see any reason why it wouldn't be, either.